src/ParticleTracer.tpp

```diff
@@ -0,0 +1,104 @@
+//
+// Created by Fuad Hasan on 2/26/25.
+//
+
+#ifndef PUMIPIC_PARTICLETRACER_H
+#define PUMIPIC_PARTICLETRACER_H
+
+#include <pumipic_adjacency.tpp>
+#include <optional>
+
+template<typename ParticleType, typename Func>
+class ParticleTracer {
+public:
+    ParticleTracer(Omega_h::Mesh &mesh, pumipic::ParticleStructure<ParticleType> *ptcls, Func &func,
+                   std::optional<double> tolerance = std::nullopt) :
+            oh_mesh_(mesh), ptcls_(ptcls), func_(func) {
+
+        elmAreas_ = Omega_h::measure_elements_real(&mesh);
+        if (tolerance.has_value()) {
+            tolerance_ = tolerance.value();
+        } else {
+            tolerance_ = pumipic::compute_tolerance_from_area(elmAreas_);
+        }
+    }
+
+    bool search() {
+        if (!validate_internal_data_sizes()) {
+            return false;
+        }
+        reset_particle_done();
+
+        auto particle_orig = ptcls_->template get<0>();
+        auto particle_dest = ptcls_->template get<1>();
+        auto particle_ids = ptcls_->template get<2>();
```

Comment:
We should provide indirection here so the user can decide which IDs correspond to the locations. This will also make it easier if we try to move the other particle arrays into the particle datastructure.

The way to do this is as follows:
1. Create a template struct called `ParticleDataStructureIndexMap` we may change name later. This should take a single template argument which will be the particle datastructure.
2. Overload the map for your particle datastructure with `constexpr int` members called `OriginalPosition`, `DestinationPosition`, `ParticleID`. For your default particle datastructure they will be set to 0, 1, 2.
3. Have your class take an additional template argument which will be this map. The default type should be `ParticleDataStructureIndexMap<ParticleType>`

Reply:
Do you mean something like this?

```c++
template<typename ParticleType>
struct ParticleDataStructureIndexMap {
    static constexpr int OriginalPosition       = 0;
    static constexpr int DestinationPosition    = 1;
    static constexpr int ParticleID             = 2;
};

template<typename ParticleType, typename Func, typename ParticleIDMap = ParticleDataStructureIndexMap<ParticleType>>
class ParticleTracer {
...
```

If this is what you mean,
1. Why does the struct need the `ParticleType` template?
1. Can't it be just an `enum class`?
2. If a user wants to specify the map themselves, they have to specify all three of them which can be difficult.

Reply:
Yes, this is what I mean. This extra indirection let's us decouple the position of the information in the particle struct from the the items. I.e., if the user wants to put the location in different positions they can easily do that.

I'm not 100% sure that we want to include the template on the particle structure since you could have the same particle structure type with different semantic meaning. But the goal there is that the compiler will "intelligently" pick the right index based on the particle structure the user passes in.